pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -2369,143 +2370,137 @@ public void resetSegment(String tableNameWithType, String segmentName, long exte
-        Map<String, String> newExternalViewStateMap = newExternalView.getStateMap(segmentToCheck);
-        if (newExternalViewStateMap == null) {
-          continue;
-        }
-        boolean allOffline = true;
-        for (String instance : instancesToCheck) {
-          if (!SegmentStateModel.OFFLINE.equals(newExternalViewStateMap.get(instance))) {
-            allOffline = false;
-            break;
-          }
-        }
-        if (allOffline) {
-          iterator.remove();
-        }
+      if (targetInstance == null || targetInstance.equals(entry.getKey())) {
+        resetPartitionAllState(entry.getKey(), tableNameWithType,
+            entry.getValue());
       }
-      Thread.sleep(EXTERNAL_VIEW_CHECK_INTERVAL_MS);
     }
-    if (!segmentInstancesToCheck.isEmpty()) {
-      throw new TimeoutException(String.format(
-          "Timed out waiting for external view to stabilize after call to disable/reset segments. "
-              + "Disable/reset might complete in the background, but skipping enable of segments of table: %s",
-          tableNameWithType));
+
+    LOGGER.info("Reset segments for table {} finished. WIth the following segments skipped: {}", tableNameWithType,
+        instanceToSkippedSegmentsMap);
+  }
+
+  /**
+   * This util is similar to {@link HelixAdmin#resetPartition(String, String, String, List)}.
+   * However instead of resetting only the ERROR state to its initial state. we reset all state regardless.
+   */
+  private void resetPartitionAllState(String instanceName, String resourceName,
+      Set<String> resetPartitionNames) {
+    LOGGER.info("Reset partitions {} for resource {} on instance {} in cluster {}.",
+        resetPartitionNames == null ? "NULL" : resetPartitionNames, resourceName,
+        instanceName, _helixClusterName);
+    HelixDataAccessor accessor = _helixZkManager.getHelixDataAccessor();
+    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+
+    // check the instance is alive
+    LiveInstance liveInstance = accessor.getProperty(keyBuilder.liveInstance(instanceName));
+    if (liveInstance == null) {
+      // check if the instance exists in the cluster
+      String instanceConfigPath = PropertyPathBuilder.instanceConfig(_helixClusterName, instanceName);
+      throw new RuntimeException(String.format("Can't find instance: %s on %s", instanceName, instanceConfigPath));
     }
 
-    // Lastly, enable segments
-    LOGGER.info("Enabling segments of table: {}", tableNameWithType);
-    for (Map.Entry<String, Set<String>> entry : instanceToResetSegmentsMap.entrySet()) {
-      _helixAdmin.enablePartition(true, _helixClusterName, entry.getKey(), tableNameWithType,
-          Lists.newArrayList(entry.getValue()));
+    // gather metadata for sending state transition message.
+    // we skip through the sanity checks normally done on Helix because in Pinot these are guaranteed to be safe.
+    IdealState idealState = accessor.getProperty(keyBuilder.idealStates(resourceName));
```

Comment:
We don't need to read ideal-state because everything extracted from the ideal state is static for pinot.

Reply:
agree. this is static per resource type. since this function is copied from helix, it is called reset partition not reset segment. we can refactor this part out later. adding todos